### ISSUE TITLE
Issue with models/segmentation/deeplabv3

### ISSUE THREAD

**Issue by ychnh** [pytorch/vision#1491]
Not sure if this class was finished implementing or not. (I see that the main class DeepLabV3 just has pass for the implementation)
However, the pyramid pooling submodule is almost functional, save for an issue with (ASPPPooling).

https://github.com/pytorch/vision/blob/1e857d93c8de081e61695dd43e6f06e3e7c2b0a2/torchvision/models/segmentation/deeplabv3.py#L53

nn.AdaptiveAvgPool2d() output size is set to 1, which changes the Height and Width of the image to 1x1 raising an error.

I don't think that this is the intended behavior of Deeplabv3 as the pooling is part of the spatial pyramid pooling designed to extract high level features along side the 3 atrous convolutions.

I think a good way resolve the issue to allow the user the specify or dynamically set as half the input Height/Width.

I could be doing something wrong so please feel free to correct me.

**Comment by fmassa** [comment]
The implementation for DeepLabV3 is finished and the pre-trained models yield similar accuracies (within a few %) as the ones reported in the original paper, and I believe the implementation is correct as it stands (but I'm happy to be proven wrong here).

From https://arxiv.org/pdf/1706.05587.pdf section 3.3

> To overcome this problem and incorporate global context
> information to the model, we adopt image-level features,
> similar to [58, 95]. Specifically, we apply global average
> pooling on the last feature map of the model,
> feed the resulting image-level features to a 1 × 1 convolution with 256
> filters (and batch normalization [38]), and then bilinearly
> upsample the feature to the desired spatial dimension

The paper mentions "global average pooling", which is in fact pooling over all the elements of the feature map.

I'm closing this issue, but let me know if you have further questions.

**Comment by ychnh** [comment]
Thank you for the reply.

You are right. I wasn't familiar with global average pooling and thought it was some type of resolution reduction pooling. But now I understand. (global avg pooling is trying to locate the logits related to num_classes)

When I was testing a model I was building, I created an input with just 1 batch, creating an error at batchnorm.
.`ValueError: Expected more than 1 value per channel when training, got input size torch.Size([1, 200, 1, 1])`

To make sure I understood this, is the following correct reasoning of why I got this error? BatchNorm2d is performed over the channels. 
Since my torch was 1,200,1,1 it only had 1 value to normalize over each channel, throwing the error.

Thanks again for the help. I'll really need to make sure to read the documentation/fine print of each method I am using without assuming what it does.
